readthedocs_ext/readthedocs.py

```diff
@@ -21,8 +24,90 @@
+    MEDIA_URL = context.get('MEDIA_URL', 'https://media.readthedocs.org/')
+
+    # Put in our media files instead of putting them in the docs.
+    for index, file in enumerate(builder.script_files):
+        if file in MEDIA_MAPPING.keys():
+            builder.script_files[index] = MEDIA_MAPPING[file] % MEDIA_URL
+            if file == "_static/jquery.js":
+                builder.script_files.insert(
+                    index + 1, "%sjavascript/jquery/jquery-migrate-1.2.1.min.js" % MEDIA_URL)
+
+
+def update_body(app, pagename, templatename, context, doctree):
+    """
+    Add Read the Docs content to Sphinx body content.
+
+    This is the most reliable way to inject our content into the page.
+    """
+
+    MEDIA_URL = context.get('MEDIA_URL', 'https://media.readthedocs.org/')
+    if app.builder.name == 'readthedocssinglehtmllocalmedia':
+        if 'html_theme' in context and context['html_theme'] == 'sphinx_rtd_theme':
+            theme_css = '_static/css/theme.css'
+        else:
+            theme_css = '_static/css/badge_only.css'
+    elif app.builder.name == 'readthedocs':
+        if 'html_theme' in context and context['html_theme'] == 'sphinx_rtd_theme':
+            theme_css = '%scss/sphinx_rtd_theme.css' % MEDIA_URL
+        else:
+            theme_css = '%scss/badge_only.css' % MEDIA_URL
+    else:
+        # Only insert on our HTML builds
+        return
+
+    template_context = context.copy()
+    template_context['theme_css'] = theme_css
+    template_context['rtd_js_url'] = '%sjavascript/readthedocs-doc-embed.js' % MEDIA_URL
+    template_context['rtd_css_url'] = '%scss/readthedocs-doc-embed.css' % MEDIA_URL
+    source = os.path.join(
+        os.path.abspath(os.path.dirname(__file__)),
+        '_templates',
+        'readthedocs-insert.html.tmpl'
+    )
+    templ = open(source).read()
+    rtd_content = app.builder.templates.render_string(templ, template_context)
+
+    # This is monkey patched on the signal because we can't know what the user
+    # has done with their `app.builder.templates` before now.
+
+    global HAS_MONKEYPATCH
+
+    if not HAS_MONKEYPATCH:
+        # Janky monkey patch of template rendering to add our content
+        old_render = app.builder.templates.render
+
+        def rtd_render(template, context):
+            """
+            A decorator that renders the content with the users template renderer,
+            then adds the Read the Docs HTML content at the end of body.
+            """
+            content = old_render(template, context)
+            end_body = content.find('</body>')
+            if end_body == -1:
+                app.warn("File doesn't look like HTML. Skipping RTD content addition")
+            # Insert our content at the end of the body.
+            content = content[:end_body] + rtd_content + content[end_body:]
+            return content
+
+        app.builder.templates.render = rtd_render
+        HAS_MONKEYPATCH = True
```

Comment:
Put what on `app`? I was saying that instead of tracking `HAS_MONKEYPATCH` in our module, we should be tracking that state on the object we're passing. The first approach above tracks the function we've patched already, though we can also do this with the second approach -- that is, something like:

```python
if not getattr(app.builder.templates.render, '_wrapped_call'):
    old_render = app.builder.templates.render
    def rtd_render(...):
        ...
        context = old_render(...)
        ...
    rtd_render._wrapped_call = old_render
    app.builder.templates.render = rtd_render
```

This is just an example, its untested.

This gives users the ability to reclaim `app.builder.templates.render._wrapped_call`, if something is completely broken with our patching. It also provides a truthier test if the object is actually patched.